plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/HOAS.hs

```diff
@@ -0,0 +1,504 @@
+    | HApply ann (HTerm name uni fun ann) (HTerm name uni fun ann)
+    | HDelay ann (HTerm name uni fun ann)
+    | HForce ann (HTerm name uni fun ann)
+    | HError ann
+
+type instance UniOf (HTerm name uni fun ann) = uni
+
+instance AsConstant (HTerm name uni fun ann) where
+    asConstant (HConstant _ val) = Just val
+    asConstant _                 = Nothing
+
+instance FromConstant (HTerm name uni fun ()) where
+    fromConstant = HConstant ()
+
+data FreeVariableError name ann = FreeVariableError ann name
+    deriving (Show, Eq)
+
+-- distributeLazily :: Functor f => (a -> f b) -> f (a -> b)
+-- distributeLazily f = f x <&> \y x' -> unsafePerformIO $ y <$ tryPutMVar hole x' where
+--     hole = unsafePerformIO newEmptyMVar
+--     x = unsafePerformIO $ readMVar hole
+
+-- -- 'distributeLazily' is overly lazy and we get shadowing handled incorrectly.
+-- toHTerm
+--     :: HasUnique name unique
+--     => Term name uni fun ann -> Either (FreeVariableError name ann) (HTerm name uni fun ann)
+-- toHTerm = go mempty where
+--     go
+--         :: HasUnique name unique
+--         => UniqueMap unique (HTerm name uni fun ann)
+--         -> Term name uni fun ann
+--         -> Either (FreeVariableError name ann) (HTerm name uni fun ann)
+--     go _   (Constant ann val)     = pure $ HConstant ann val
+--     go _   (Builtin ann fun)      = pure $ HBuiltin ann fun
+--     go env (Var ann name)         =
+--         case lookupName name env of
+--             Nothing   -> Left $ FreeVariableError ann name
+--             Just term -> Right term
+--     go env (LamAbs ann name body) =
+--         -- That would require an explicit @Lazy@ wrapper, if we didn't use lazy @IntMap@s, I guess.
+--         HLamAbs ann name <$> distributeLazily (\var -> go (insertByName name var env) body)
+--     go env (Apply ann fun arg)    = HApply ann <$> go env fun <*> go env arg
+--     go env (Delay ann term)       = HDelay ann <$> go env term
+--     go env (Force ann term)       = HForce ann <$> go env term
+--     go _   (Error ann)            = pure (HError ann)
+
+toHTerm
+    :: HasUnique name unique
+    => Term name uni fun ann -> Either (FreeVariableError name ann) (HTerm name uni fun ann)
+toHTerm = pure . go mempty where
+    go
+        :: HasUnique name unique
+        => UniqueMap unique (HTerm name uni fun ann)
+        -> Term name uni fun ann
+        -> HTerm name uni fun ann
+    go _   (Constant ann val)     = HConstant ann val
+    go _   (Builtin ann fun)      = HBuiltin ann fun
+    go env (Var ann name)         =
+        case lookupName name env of
+            Nothing   -> error "free variable" -- throw $ FreeVariableError ann name
+            Just term -> term
+    go env (LamAbs ann name body) =
+        -- That would require an explicit @Lazy@ wrapper, if we didn't use lazy @IntMap@s, I guess.
+        HLamAbs ann name $ \var -> go (insertByName name var env) body
```

Comment:
The body of the function explicitly uses the environment to map the unique to the argument value, and this will happen every time the function is called. So we have these unique->term environments around at runtime, and we use them every time we do a variable lookup. So in that respect this is still doing a lot of the same work as the CEK machine at runtime! I wonder if this explains why the performance isn't as good as we expected.

I'm not really sure what the alternative is. I think I was deceived by the examples where you write HOAS terms directly: then you don't need environments and you can just use GHC's handling of variables. But I can't see how to do that for an arbitrary term at runtime without generating Haskell and then compiling that...

Reply:
> So we have these unique->term environments around at runtime, and we use them every time we do a variable lookup.

Not every time. Only once when we convert the first-order AST into the higher-order one. After that environment lookups are not used and we simply use meta functions for doing substitutions for us.

So yes, there's a cost to this "initial" conversion (the quotes are due to the fact that conversion is interleaved with evaluation, but morally, it's initial), but it seems inevitable as you say.

Reply:
Maybe I'm missing something: `HLamAbs` contains a function, and the body of that function uses the environment to do lookups. So every time you call that function, you'll be doing something with the environments, no?

Reply:
Hm. It sounds like you're right and I'm wrong. I'll think more about this later. Thanks for bringing that up.

Reply:
So yeah, you're right. Oh, I should mention that issue in docs.